pkg/mirror/transport.go

```diff
@@ -0,0 +1,257 @@
+package mirror
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/TierMobility/boring-registry/pkg/auth"
+	"github.com/TierMobility/boring-registry/pkg/core"
+
+	"github.com/go-kit/kit/endpoint"
+	httptransport "github.com/go-kit/kit/transport/http"
+	"github.com/gorilla/mux"
+	"github.com/pkg/errors"
+)
+
+type muxVar string
+
+const (
+	varHostname     muxVar = "hostname"
+	varNamespace    muxVar = "namespace"
+	varName         muxVar = "name"
+	varVersion      muxVar = "version"
+	varOS           muxVar = "os"
+	varArchitecture muxVar = "architecture"
+)
+
+type header string
+
+// MakeHandler returns a fully initialized http.Handler.
+func MakeHandler(svc Service, auth endpoint.Middleware, options ...httptransport.ServerOption) http.Handler {
+	r := mux.NewRouter().StrictSlash(true)
+
+	r.Methods("GET").Path(`/{hostname}/{namespace}/{name}/index.json`).Handler(
+		httptransport.NewServer(
+			auth(listProviderVersionsEndpoint(svc)),
+			decodeListVersionsRequest,
+			EncodeJSONResponse,
+			append(
+				options,
+				httptransport.ServerBefore(extractMuxVars(varHostname, varNamespace, varName)),
+				httptransport.ServerBefore(extractHeaders("Authorization")),
+			)...,
+		),
+	)
+
+	r.Methods("GET").Path(`/{hostname}/{namespace}/{name}/{version}.json`).Handler(
+		httptransport.NewServer(
+			auth(listProviderInstallationEndpoint(svc)),
+			decodeListInstallationRequest,
+			EncodeJSONResponse,
+			append(
+				options,
+				httptransport.ServerBefore(extractMuxVars(varHostname, varNamespace, varName, varVersion)),
+				httptransport.ServerBefore(extractHeaders("Authorization")),
+			)...,
+		),
+	)
+
+	r.Methods("GET").Path(`/{hostname}/{namespace}/{name}/terraform-provider-{nameplaceholder}_{version}_{os}_{architecture}.zip`).Handler(
+		httptransport.NewServer(
+			auth(retrieveProviderArchiveEndpoint(svc)),
```

Comment:
I guess this endpoint is secured with a token. See https://developer.hashicorp.com/terraform/internals/provider-network-mirror-protocol#authentication

> Terraform does not send credentials when retrieving the archives whose URLs are given in the "List Available Installation Packages" response below. If a particular mirror considers the distribution packages themselves to be sensitive then it must use cryptographically-secure, user-specific, and time-limited URLs in the metadata response. Strategies for doing so are out of scope of this protocol documentation.

Reply:
Good catch! This is kind of a problem for my approach that I've chosen. One possible solution would be to remove this endpoint entirely. Instead, we return the upstream download URL to the client in case the provider is not mirrored yet. In case it is mirrored, we return the pre-signed url to the object in the mirror.

Reply:
In case authorization is configured we're now returning download URLs with a `token=<token>` query parameter since 8ad6a194b2246c5dccd13c18a9f3a136b792f8a2 . The `<token>` is either a static token or an oauth2 token, depending on the client and boring-registry's configuration.
In case the registry is unable to fetch the provider information from the upstream registry, we directly return the pre-signed URLs for the providers in the storage, so we're avoiding one more round-trip